lib/optimizer/src/ImageDimensions.php

```diff
@@ -0,0 +1,226 @@
+<?php
+
+namespace AmpProject\Optimizer;
+
+use AmpProject\Attribute;
+use AmpProject\Layout;
+use DOMElement;
+
+final class ImageDimensions
+{
+
+    /**
+     * Images smaller than 150px are considered tiny.
+     *
+     * @var int
+     */
+    const TINY_THRESHOLD = 150;
+
+    /**
+     * Image that this represents the dimensions of.
+     *
+     * @var DOMElement
+     */
+    private $image;
+
+    /**
+     * Width of the image.
+     *
+     * @var int|float|string|null
+     */
+    private $width;
+
+    /**
+     * Height of the image.
+     *
+     * @var int|float|string|null
+     */
+    private $height;
+
+    /**
+     * Layout of the image.
+     *
+     * @var string|null
+     */
+    private $layout;
+
+    /**
+     * ImageDimensions constructor.
+     *
+     * @param DOMElement $image Image to represent the dimensions of.
+     */
+    public function __construct(DOMElement $image)
+    {
+        $this->image = $image;
+    }
+
+    /**
+     * Get the dimensions to use from an element's parent(s).
+     *
+     * @return int[] Array containing the width and the height.
+     */
+    public function getDimensionsFromParent()
+    {
+        $level   = 0;
+        $element = $this->image;
+        while ($element->parentNode && ++$level < 3) {
+            $element = $element->parentNode;
+
+            if (! $element instanceof DOMElement) {
+                continue;
+            }
+
+            $width = $element->hasAttribute(Attribute::WIDTH)
+                ? $element->getAttribute(Attribute::WIDTH)
+                : -1;
+
+            $height = $element->hasAttribute(Attribute::HEIGHT)
+                ? $element->getAttribute(Attribute::HEIGHT)
+                : -1;
+
+            if (empty($width)) {
+                $width = -1;
+            }
+
+            if (empty($height)) {
+                $height = -1;
+            }
+
+            // Skip elements that don't provide any dimensions.
+            if ($width === -1 && $height === -1) {
+                continue;
+            }
+
+            return [(int)$width, (int)$height];
```

Comment:
According to the docs, the `width` & `height` attributes would be an "integer pixel value": https://amp.dev/documentation/guides-and-tutorials/learn/amp-html-layout/?format=websites

Do you have a source you can link to re. attached units and/or float values?

Reply:
Even though the docs say they should be integer pixel values, this is not enforced and non-integers can be supplied.

For floats, see: https://github.com/ampproject/amphtml/issues/27528

And [this](https://playground.amp.dev/#share=PCFkb2N0eXBlIGh0bWw+CjxodG1sIOKaoT4KPGhlYWQ+CiAgPG1ldGEgY2hhcnNldD0idXRmLTgiPgogIDx0aXRsZT5NeSBBTVAgUGFnZTwvdGl0bGU+CiAgPGxpbmsgcmVsPSJjYW5vbmljYWwiIGhyZWY9InNlbGYuaHRtbCIgLz4KICA8bWV0YSBuYW1lPSJ2aWV3cG9ydCIgY29udGVudD0id2lkdGg9ZGV2aWNlLXdpZHRoLG1pbmltdW0tc2NhbGU9MSxpbml0aWFsLXNjYWxlPTEiPgogIDxzdHlsZSBhbXAtYm9pbGVycGxhdGU+Ym9keXstd2Via2l0LWFuaW1hdGlvbjotYW1wLXN0YXJ0IDhzIHN0ZXBzKDEsZW5kKSAwcyAxIG5vcm1hbCBib3RoOy1tb3otYW5pbWF0aW9uOi1hbXAtc3RhcnQgOHMgc3RlcHMoMSxlbmQpIDBzIDEgbm9ybWFsIGJvdGg7LW1zLWFuaW1hdGlvbjotYW1wLXN0YXJ0IDhzIHN0ZXBzKDEsZW5kKSAwcyAxIG5vcm1hbCBib3RoO2FuaW1hdGlvbjotYW1wLXN0YXJ0IDhzIHN0ZXBzKDEsZW5kKSAwcyAxIG5vcm1hbCBib3RofUAtd2Via2l0LWtleWZyYW1lcyAtYW1wLXN0YXJ0e2Zyb217dmlzaWJpbGl0eTpoaWRkZW59dG97dmlzaWJpbGl0eTp2aXNpYmxlfX1ALW1vei1rZXlmcmFtZXMgLWFtcC1zdGFydHtmcm9te3Zpc2liaWxpdHk6aGlkZGVufXRve3Zpc2liaWxpdHk6dmlzaWJsZX19QC1tcy1rZXlmcmFtZXMgLWFtcC1zdGFydHtmcm9te3Zpc2liaWxpdHk6aGlkZGVufXRve3Zpc2liaWxpdHk6dmlzaWJsZX19QC1vLWtleWZyYW1lcyAtYW1wLXN0YXJ0e2Zyb217dmlzaWJpbGl0eTpoaWRkZW59dG97dmlzaWJpbGl0eTp2aXNpYmxlfX1Aa2V5ZnJhbWVzIC1hbXAtc3RhcnR7ZnJvbXt2aXNpYmlsaXR5OmhpZGRlbn10b3t2aXNpYmlsaXR5OnZpc2libGV9fTwvc3R5bGU+PG5vc2NyaXB0PjxzdHlsZSBhbXAtYm9pbGVycGxhdGU+Ym9keXstd2Via2l0LWFuaW1hdGlvbjpub25lOy1tb3otYW5pbWF0aW9uOm5vbmU7LW1zLWFuaW1hdGlvbjpub25lO2FuaW1hdGlvbjpub25lfTwvc3R5bGU+PC9ub3NjcmlwdD4KICA8c2NyaXB0IGFzeW5jIHNyYz0iaHR0cHM6Ly9jZG4uYW1wcHJvamVjdC5vcmcvdjAuanMiPjwvc2NyaXB0PgogIDxzdHlsZSBhbXAtY3VzdG9tPgogICAgaDEgewogICAgICBtYXJnaW46IDFyZW07CiAgICB9CiAgPC9zdHlsZT4KPC9oZWFkPgo8Ym9keT4KPGFtcC1pbWcgc3JjPSJodHRwczovL2FtcC5kZXYvc3RhdGljL2ltZy9pY29ucy9pY29uLTUxMng1MTIucG5nIiB3aWR0aD0iNTB2dyIgaGVpZ2h0PSI1MHZoIj48L2FtcC1pbWc+CjwvYm9keT4KPC9odG1sPgo=) is valid AMP:

```html
<amp-img src="https://amp.dev/static/img/icons/icon-512x512.png" width="50vw" height="50vh"></amp-img>
```

The use of decimals and/or CSS units been allowed accidentally, but since it is valid it cannot be removed.

Reply:
I will do this in a separate PR, to not block this one further and move it over to the separate repo.

Reply:
=> https://github.com/ampproject/amp-toolbox-php/issues/4